metagraph/tests/algorithms/test_embedding.py

```diff
@@ -130,56 +170,363 @@ def cmp_func(embedding):
+
+        assert a_variances.max() < 0.20
+        assert b_variances.max() < 0.20
+
+    MultiVerify(dpr).compute(
+        "embedding.train.hope.katz", graph, embedding_size, beta,
+    ).normalize(
+        (dpr.types.Matrix.NumpyMatrixType, dpr.types.NodeMap.NumpyNodeMapType)
+    ).custom_compare(
         cmp_func
     )
 
 
-# def test_hope_katz(default_plugin_resolver):
-#     dpr = default_plugin_resolver
+def test_graph_sage_mean(default_plugin_resolver):
+    """
+== Training Subgraph ==
+
+[Training Subgraph A (fully connected), nodes 0..9] --------------|
+                           |                                      |
+                    node 9999_09_10                               |
+                           |                                      |
+[Training Subgraph B (fully connected), nodes 10..19]     node 9999_29_00
+                           |                                      |
+                    node 9999_19_20                               |
+                           |                                      |
+[Training Subgraph C (fully connected), nodes 10..19] -------------
+
+Training Subgraph A nodes all have feature vector [1, 0, 0, 0, 0, ..., 0]
+Training Subgraph B nodes all have feature vector [0, 1, 0, 0, 0, ..., 0]
+Training Subgraph C nodes all have feature vector [0, 0, 1, 0, 0, ..., 0]
+Nodes 9999_09_10, 9999_19_20, and node 9999_29_00 have the zero vector as their node features.
+
+
+
+== Testing Subgraph ==
+
+[Testing Subgraph A (fully connected), nodes 8888_00..8888_19]
+                        |
+                 node 8888_00_20
+                        |
+[Testing Subgraph B (fully connected), nodes 8888_20..8888_49]
+
+Testing Subgraph A nodes all have feature vector [1, 0, 0, 0, 0, ..., 0]
+Testing Subgraph B nodes all have feature vector [0, 1, 0, 0, 0, ..., 0]
+Node 8888_00_20 hsa the zero vector as a its node features.
 
-#     # make two unidirectional circle graphs connected by one node
-#     a_nodes = np.arange(10)
-#     b_nodes = np.arange(10, 20)
 
-#     nx_graph = nx.DiGraph()
 
-#     for i in range(9):
-#         a_src_node = a_nodes[i]
-#         a_dst_node = a_nodes[i+1]
-#         b_src_node = b_nodes[i]
-#         b_dst_node = b_nodes[i+1]
-#         nx_graph.add_edge(a_src_node, a_dst_node, weight = 25)
-#         nx_graph.add_edge(b_src_node, b_dst_node, weight = 25)
+== Differences Between Training & Testing Graphs ==
 
-#     nx_graph.add_edge(0, 1_000, weight = 1)
-#     nx_graph.add_edge(10, 1_000, weight = 1)
+All the complete subgraphs in the training graph have 10 nodes, but the complete subgraphs in the testing graph do NOT.
 
-#     graph = dpr.wrappers.Graph.NetworkXGraph(nx_graph)
+The test verifies for the testing graph that the 20 nearest neighbors in the embedding space of each node are all part of the same complete subgraph.
+    """
 
-#     mv = MultiVerify(dpr)
+    try:
+        from sklearn.neighbors import NearestNeighbors
+    except:
```

Comment:
`except ImportError:`

Reply:
https://github.com/metagraph-dev/metagraph/pull/138/commits/c6f88dfd8b708da632b8f4336400a510ed49f371 makes it so that it catches `ModuleNotFoundError`.